crud/select/plan.lua

```diff
@@ -226,7 +228,14 @@ function select_plan.new(space, conditions, opts)
         end
     end
 
-    local sharding_index = primary_index -- XXX: only sharding by primary key is supported
+    local ddl_sharding_key = sharding.get_ddl_sharding_key(space_name, box)
+    local sharding_index = primary_index
+    if ddl_sharding_key ~= nil then
+        sharding_index, err = utils.get_sharding_key_fieldnos(space_format, space_name, ddl_sharding_key)
+        if err ~= nil then
+            return nil, err
+        end
+    end
 
     -- get sharding key value
     local sharding_key
```

Comment:
Several lines below:

```lua
    if sharding_key ~= nil and opts.force_map_call ~= true then
        total_tuples_count = 1
        scan_iter = box.index.REQ
    end
```

It assumes an index as unique. It was true for the primary key, but it is not guaranteed for a sharding key. Please, test this case.

Reply:
Added:
```
    ddl_sharding_key[vinyl].test_non_unique_index ... (0.013s) Ok
    ddl_sharding_key[memtx].test_non_unique_index ... (0.013s) Ok
```

Reply:
But you requested just one tuple in this test. Of course, it'll not be broken by `total_tuples_count = 1`.

Reply:
With two Ivan's Bunin in a table CRUD selects both of them.

```lua
pgroup.test_non_unique_index = function(g)                                       
    local space_name = 'customers_name_key_non_uniq_index'                       
    local customers = helpers.insert_objects(g, space_name, {                    
        {id = 1, name = 'Viktor Pelevin', age = 58},                             
        {id = 2, name = 'Isaac Asimov', age = 72},                               
        {id = 3, name = 'Aleksandr Solzhenitsyn', age = 89},                     
        {id = 4, name = 'James Joyce', age = 59},                                
        {id = 5, name = 'Oscar Wilde', age = 46},                                
        {id = 6, name = 'Ivan Bunin', age = 83},                                 
        {id = 7, name = 'Ivan Turgenev', age = 64},                              
        {id = 8, name = 'Alexander Ostrovsky', age = 63},                        
        {id = 9, name = 'Anton Chekhov', age = 44},                              
        {id = 10, name = 'Ivan Bunin', age = 83},                                
    })                                                                           
    t.assert_equals(#customers, 10)                                              
                                                                                 
    local result, err = g.cluster.main_server.net_box:call('crud.select', {         
        space_name, {{'==', 'name', 'Ivan Bunin'}}                               
    })                                                                           
    t.assert_equals(err, nil)                                                    
    t.assert_not_equals(result, nil)                                             
    t.assert_equals(#result.rows, 2)                                             
end  
```

Reply:
Thanks, I know about it and we discussed it verbally.

Reply:
Fixed it in commit 'Support custom sharding key in select'.